android/jni/com/mapswithme/util/HttpClient.cpp

```diff
@@ -0,0 +1,303 @@
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*******************************************************************************/
+#include <jni.h>
+
+#include "../core/jni_helper.hpp"
+
+#include "platform/http_client.hpp"
+
+#include "base/logging.hpp"
+#include "base/assert.hpp"
+
+#include "std/string.hpp"
+
+namespace
+{
+template <typename T, typename D>
+unique_ptr<T, D> MakeScopedPointer(T * ptr, D deleter)
+{
+  return unique_ptr<T, D>(ptr, deleter);
+}
+
+// Scoped environment which can attach to any thread and automatically detach
+class ScopedEnv final
+{
+public:
+  ScopedEnv(JavaVM * vm)
+  {
+    JNIEnv * env;
+    auto result = vm->GetEnv(reinterpret_cast<void **>(&env), JNI_VERSION_1_6);
+    if (result == JNI_EDETACHED)
+    {
+      result = vm->AttachCurrentThread(&env, nullptr);
+      m_needToDetach = (result == JNI_OK);
+    }
+
+    if (result == JNI_OK)
+    {
+      m_env = env;
+      m_vm = vm;
+    }
+  }
+
+  ~ScopedEnv()
+  {
+    if (m_vm != nullptr && m_needToDetach)
+      m_vm->DetachCurrentThread();
+  }
+
+  JNIEnv * operator->() { return m_env; }
+  operator bool() const { return m_env != nullptr; }
+  JNIEnv * get() { return m_env; }
+
+private:
+  bool m_needToDetach = false;
+  JNIEnv * m_env = nullptr;
+  JavaVM * m_vm = nullptr;
+};
+}  // namespace
+
+#define CLEAR_AND_RETURN_FALSE_ON_EXCEPTION \
+  if (env->ExceptionCheck()) {              \
+    env->ExceptionDescribe();               \
+    env->ExceptionClear();                  \
+    return false;                           \
+  }
+
+//***********************************************************************
+// Exported functions implementation
+//***********************************************************************
+namespace platform
+{
+bool HttpClient::RunHttpRequest()
+{
+  ScopedEnv env(jni::GetJVM());
+
+  if (!env)
+    return false;
+
+  auto const deleter = [&env](jobject o) { env->DeleteLocalRef(o); };
+
+  // Create and fill request params.
+  auto const jniUrl = MakeScopedPointer(jni::ToJavaString(env.get(), m_urlRequested), deleter);
```

Comment:
Why do we need such ugly jni code? I guess it's better to convert and pass arguments to a java method, and then convert its result back.